cumulus/client/consensus/aura/src/collators/mod.rs

```diff
@@ -243,3 +250,110 @@ where
+
+		let mut block_import_params = BlockImportParams::new(origin, header);
+		block_import_params.fork_choice = Some(ForkChoiceStrategy::Custom(import_as_best));
+		block_import_params.body = Some(body);
+		importer.import_block(block_import_params).await.unwrap();
+	}
+
+	fn sproof_with_parent_by_hash(client: &Client, hash: PHash) -> RelayStateSproofBuilder {
+		let header = client.header(hash).ok().flatten().expect("No header for parent block");
+		let included = HeadData(header.encode());
+		let mut x = RelayStateSproofBuilder::default();
+		x.para_id = cumulus_test_client::runtime::PARACHAIN_ID.into();
+		x.included_para_head = Some(included);
+
+		x
+	}
+	async fn build_and_import_block(client: &Client, included: Hash) -> Block {
+		let sproof = sproof_with_parent_by_hash(client, included);
+
+		let block_builder = client.init_block_builder(None, sproof).block_builder;
+
+		let block = block_builder.build().unwrap().block;
+
+		let origin = BlockOrigin::NetworkInitialSync;
+		import_block(client, block.clone(), origin, true).await;
+		block
+	}
+
+	fn set_up_components() -> (Arc<Client>, KeystorePtr) {
+		let keystore = Arc::new(sp_keystore::testing::MemoryKeystore::new()) as Arc<_>;
+		for key in sp_keyring::Sr25519Keyring::iter() {
+			Keystore::sr25519_generate_new(
+				&*keystore,
+				sp_application_crypto::key_types::AURA,
+				Some(&key.to_seed()),
+			)
+			.expect("Can insert key into MemoryKeyStore");
+		}
+		(Arc::new(TestClientBuilder::new().build()), keystore)
+	}
+
+	/// This tests a special scenario where the unincluded segment in the runtime
+	/// is full. We are calling `can_build_upon`, passing the last built block as the
+	/// included one. In the runtime we will not find the hash of the included block in the
+	/// unincluded segment. The `can_build_upon` runtime API would therefore return `false`, but
+	/// we are ensuring on the node side that we are are always able to build on the included block.
+	#[tokio::test]
+	async fn test_can_build_upon() {
+		let (client, keystore) = set_up_components();
+
+		let genesis_hash = client.chain_info().genesis_hash;
+		let mut last_hash = None;
+
+		// Fill up the unincluded segment tracker in the runtime.
+		for _ in 0..cumulus_test_client::runtime::UNINCLUDED_SEGMENT_CAPACITY {
+			let block = build_and_import_block(&client, genesis_hash).await;
+			last_hash = Some(block.header().hash());
+		}
+
+		let last_hash = last_hash.expect("must exist");
+		// We don't care much about the slots.
+		let para_slot = Slot::from(u64::MAX);
+		let relay_slot = Slot::from(u64::MAX);
+
+		// Blocks were built with the genesis hash set as included block.
+		// We call `can_build_upon` with the last built block as the included block.
+		let result = can_build_upon::<_, _, sp_consensus_aura::sr25519::AuthorityPair>(
+			para_slot,
+			relay_slot,
+			Timestamp::default(),
+			last_hash,
+			last_hash,
+			&*client,
+			&keystore,
+		)
+		.await;
+		assert!(result.is_some());
```

Comment:
IMO it would be better to write the test roughly like this:

```
while can_build_upon(genesis_hash_as_included_block, last_block) { build_new_block() }

can_build_upon(last_block, last_block)
```

This way we know that `can_build_upon` returned `false` before and returns `true` when we switch the included block.